src/payment/coupon/CouponForm.jsx

```diff
@@ -33,16 +33,69 @@ export class CouponForm extends Component {
     this.props.removeCoupon(this.props.voucherId);
   }
 
+  renderInvalidMessage() {
+    const { code, errorCode, intl } = this.props;
+    if (errorCode === null) {
+      return null;
+    }
+    let invalidMessage = null;
+    switch (errorCode) {
+      case 'empty_basket':
+        invalidMessage = intl.formatMessage(messages['payment.coupon.error.empty_basket'], {
```

Comment:
I'd probably either make a map from error code to message id, or do a shorter version of the case:

`
case 'empty_basket':
case 'too_many_kittens':
case 'missing_cute_bow':
    messageId = 'payment.coupon.error.' + errorCode; // not going to try to mess with Markdown and backticks in this comment!
    break;
default:
    messageId = 'payment.coupon.error.doh';
}

return intl.formatMessage(messages[messageId, { code });
`